File: backend/src/main/kotlin/com/alpha/domain/repository/SessionRepository.kt
package com.alpha.domain.repository

import com.alpha.domain.entity.SessionEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.time.OffsetDateTime
import java.util.*

interface SessionRepository : JpaRepository<SessionEntity, UUID> {
    fun findByToken(token: String): SessionEntity?
    fun findByRefreshToken(refreshToken: String): SessionEntity?
    @Query("SELECT s FROM SessionEntity s WHERE s.user.id = :userId")
    fun findByUserId(@Param("userId") userId: UUID): List<SessionEntity>
    
    @Query("SELECT s FROM SessionEntity s WHERE s.expiresAt < :now")
    fun findExpiredSessions(@Param("now") now: OffsetDateTime): List<SessionEntity>
    
    @Modifying
    @Query("DELETE FROM SessionEntity s WHERE s.expiresAt < :now")
    fun deleteExpiredSessions(@Param("now") now: OffsetDateTime): Int
    
    @Modifying
    @Query("DELETE FROM SessionEntity s WHERE s.user.id = :userId")
    fun deleteByUserId(@Param("userId") userId: UUID): Int
}

File: backend-scala-zio/src/main/scala/com/alpha/repository/SessionRepository.scala
package com.alpha.repository

import zio.*
import com.alpha.domain.model.UserSession
import java.util.UUID
import io.getquill.*
import io.getquill.extras.OffsetDateTimeOps

trait SessionRepository:
  def findById(id: UUID): Task[Option[UserSession]]
  def findByUserId(userId: UUID): Task[List[UserSession]]
  def findByRefreshToken(token: String): Task[Option[UserSession]]
  def findByToken(token: String): Task[Option[UserSession]]
  def create(session: UserSession): Task[UUID]
  def revoke(id: UUID): Task[Int]
  def revokeByUserId(userId: UUID): Task[Int]
  def deleteExpired: Task[Int]

object SessionRepository:
  val layer: ZLayer[PostgresCtx, Nothing, SessionRepository] = 
    ZLayer.fromFunction(new SessionRepositoryImpl(_))

class SessionRepositoryImpl(ctx: PostgresCtx) extends SessionRepository:

  import ctx.*

  override def findById(id: UUID): Task[Option[UserSession]] = 
    ZIO.attempt:
      run(query[UserSession].filter(_.id == lift(id))).headOption

  override def findByUserId(userId: UUID): Task[List[UserSession]] = 
    ZIO.attempt:
      run(query[UserSession].filter(s => s.userId == lift(userId) && s.isRevoked == lift(false))).toList

  override def findByRefreshToken(token: String): Task[Option[UserSession]] = 
    ZIO.attempt:
      run(query[UserSession].filter(_.refreshToken == lift(token))).headOption

  override def findByToken(token: String): Task[Option[UserSession]] = 
    ZIO.attempt:
      run(query[UserSession].filter(_.token == lift(token))).headOption

  override def create(session: UserSession): Task[UUID] = 
    ZIO.attempt:
      run(query[UserSession].insertValue(lift(session)))
      session.id

  override def revoke(id: UUID): Task[Int] = 
    ZIO.attempt:
      run(query[UserSession]
        .filter(_.id == lift(id))
        .update(_.isRevoked -> lift(true)))
      1

  override def revokeByUserId(userId: UUID): Task[Int] = 
    ZIO.attempt:
      run(query[UserSession]
        .filter(_.userId == lift(userId))
        .update(_.isRevoked -> lift(true)))
      1

  override def deleteExpired: Task[Int] = 
    ZIO.attempt:
      run(query[UserSession]
        .filter(_.expiresAt < lift(java.time.OffsetDateTime.now()))
        .delete)
      0
